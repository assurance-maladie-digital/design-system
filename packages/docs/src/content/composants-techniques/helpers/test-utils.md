---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
