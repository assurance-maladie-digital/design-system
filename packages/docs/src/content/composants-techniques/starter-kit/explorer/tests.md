Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
