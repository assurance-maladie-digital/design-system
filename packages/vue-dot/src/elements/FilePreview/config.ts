export const config = {
	pdf: {
		height: '556px',
		width: '100%'
	},
	image: {
		width: '100%'
	}
};
