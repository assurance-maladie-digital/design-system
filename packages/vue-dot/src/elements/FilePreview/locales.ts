export const locales = {
	previewNotAvailable: 'Impossible de prévisualiser le fichier.',
	previewTypeNotAvailable: 'Impossible de prévisualiser ce type de fichier.'
};
