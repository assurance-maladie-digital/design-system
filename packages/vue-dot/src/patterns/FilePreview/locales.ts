export const locales = {
	previewNotAvailable: 'Impossible de pr√©visualiser le fichier'
};
