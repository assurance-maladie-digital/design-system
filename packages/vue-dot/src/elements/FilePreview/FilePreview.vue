<template>
	<div
		v-if="file"
		class="vd-file-preview"
	>
		<object
			v-if="isPdf"
			:data="fileURL"
			v-bind="filePreviewOptions.pdf"
			type="application/pdf"
			@load="revokeFileURL"
		>
			<p class="mb-0">
				{{ locales.previewNotAvailable }}
			</p>
		</object>

		<img
			v-else-if="isImage"
			:src="fileURL"
			:alt="filePreviewOptions.image.alt || ''"
			v-bind="filePreviewOptions.image"
			@load="revokeFileURL"
		>

		<slot v-else>
			<p class="mb-0">
				{{ locales.previewTypeNotAvailable }}
			</p>
		</slot>
	</div>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';

	import { Options } from '../../mixins/customizable';

	import deepMerge from 'deepmerge';

	const Props = Vue.extend({
		props: {
			file: {
				// File is not a valid prop type,
				// use null to allow any type & provide custom validation
				type: null as unknown as PropType<File | Blob>,
				default: null,
				/** @see https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VFileInput/VFileInput.ts#L71 */
				validator(value): boolean {
					return typeof value === 'object';
				}
			},
			options: {
				type: Object as PropType<Options>,
				default: () => ({})
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component<FilePreview>({
		watch: {
			file: {
				handler(): void {
					this.getFileURL();
				},
				immediate: true
			}
		}
	})
	export default class FilePreview extends MixinsDeclaration {
		locales = locales;
		fileURL = '';

		get filePreviewOptions(): Options {
			return deepMerge<Options>(config, this.options);
		}

		get isPdf(): boolean {
			return this.file?.type === 'application/pdf';
		}

		get isImage(): boolean {
			return this.file ? /^image\/.*$/.test(this.file.type) : false;
		}

		getFileURL(): void {
			if (!this.file) {
				return;
			}

			this.fileURL = window.URL.createObjectURL(this.file);
		}

		revokeFileURL(): void {
			URL.revokeObjectURL(this.fileURL);
		}
	}
</script>
