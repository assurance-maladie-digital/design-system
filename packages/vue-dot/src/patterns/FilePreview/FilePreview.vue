<template>
	<object
		v-if="isFilePdf"
		:data="fileSrcComputed"
		v-bind="$attrs.pdf"
		type="application/pdf"
		@load="revokeFileSrc"
	>
		<p>{{ locales.previewNotAvailable }}</p>
	</object>
	<img
		v-else-if="isFileImage"
		:src="fileSrcComputed"
		v-bind="$attrs.image"
		@load="revokeFileSrc"
	>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { locales } from './locales';

	const Props = Vue.extend({
		props: {
			file: {
				type: null as unknown as PropType<File>,
				default: undefined
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component
	export default class FilePreview extends MixinsDeclaration {
		locales = locales;
		fileSrc = '';

		get fileSrcComputed(): string {
			if (this.file) {
				this.fileSrc = URL.createObjectURL(this.file);
			}
			return this.fileSrc;
		}

		get isFilePdf(): boolean {
			return this.file?.type === 'application/pdf';
		}

		get isFileImage(): boolean {
			return this.file ? /^image\/.*$/.test(this.file.type) : false;
		}

		revokeFileSrc(): void {
			if (this.file) {
				URL.revokeObjectURL(this.fileSrc);
			}
		}
	}
</script>
