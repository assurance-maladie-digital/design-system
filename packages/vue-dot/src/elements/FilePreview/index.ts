import FilePreview from './FilePreview.vue';

export default FilePreview;
