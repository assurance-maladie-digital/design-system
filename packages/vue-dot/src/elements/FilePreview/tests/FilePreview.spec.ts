import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import FilePreview from '../';

let wrapper: Wrapper<Vue>;

const testFile = {
	name: 'avatar.png',
	size: 1000,
	type: 'image/png'
} as File;

describe('FilePreview', () => {
	global.URL.createObjectURL = jest.fn();

	it('renders correctly', () => {
		wrapper = mountComponent(FilePreview, {
			propsData: {
				file: testFile
			}
		});

		expect(wrapper).toMatchSnapshot();
	});
});
