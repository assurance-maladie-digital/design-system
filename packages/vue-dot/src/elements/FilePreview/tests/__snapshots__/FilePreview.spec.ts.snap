// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilePreview renders correctly 1`] = `<div class="vd-file-preview"><img alt="" width="100%"></div>`;
